0000-improved-flow-types-for-babel-types.md

```diff
@@ -0,0 +1,164 @@
+
+
+# Detailed design
+
+The proposal of this RFC is to change today's typing to:
+
+* Use exact types instead of `class`es to make it clear that `@babel/types` does not export values for node types.
+* Use a union type for `BabelNode` rather than a base class. This allows to refine on node types using `node.type === 'Identifier'` or paths `path.node.type === 'Identifier'`.
+
+Example Usage:
+
+```
+import * as t from '@babel/types';
+import type {StringLiteral, NumericLiteral} from '@babel/types';
+
+function asStringLiteral(ast: ?StringLiteral | NumericLiteral): BabelNodeStringLiteral | null {
+  if (ast == null) {
+      return t.stringLiteral("null");
+  }
+
+  if (ast.type === "StringLiteral") {
+    return ast;
+  }
+
+  return t.stringLiteral(String(ast.value));
+}
+```
+
+This changes can be implemented by changing the type definition generation script in `babel/types` for flow only. 
+
+## Regression
+
+Flow does not support a `%check refinement` for types or interfaces. Typing the `is*` methods to refine the `node` type will, therefore, no longer possible.
+
+# Drawbacks
+
+* Braking Change: Applications that use the existing flow types and rely on the `types.is*` methods would have to migrate to `node.type ===` checks.
+* Unions can potentially be expensive during type checking. Flow does some optimisations and I tried out the my definition file and haven't noticed a regression compared to the existing type definition. 
+* Union type error messages can be hard to understand because flow lists each possible option.
+
+# Alternatives
+
+## Keep as is
+
+The drawbacks are:
+
+* Having to import the `is*` methods for type refinement to work
+* Defining the node types as classes is misguiding for users that the node types are modelled as values (classes) when they're not
+* Type refinements on `path` using `is*` or `path.type === 'Identifier'` are not supported. 
+
+## Inexact vs Exact Types
+
+An alternative is to define the node types either as inexact types or interfaces. The difference is this would allow consumers to set additional props on the nodes. 
+
+I believe that exact types are the right choice here since additional properties are e.g. not supported by `babel/generate`. 
+
+# Adoption strategy
+
+There are probably a couple of options but one would be that:
+
+1. Publish the old flow types on [flow-typed](https://github.com/flow-typed/flow-typed) so that users can ping the version
+2. Update the type definition in `babel/types`. 
+
+There's still disturbance since consumers will have to upgrade at some point if they want to use newly added node types. 
+
+# How we teach this
+
+I don't believe there's any teaching needed nor any need to re-organize the documentation
+
+# Open questions
+
+* Should the global `BabelNode*` exports be replaced by module level exports without the `BabelNode` prefix to better align with the type script types or should the definition file export both until the next major version.
```

Comment:
We can both use the global `BabelNode*` version and normal named exports for now, and then remove the global version in Babel 8.